File: backend/GameStoreAPI/Queries/GameListQuery.cs
using GameStoreAPI.Models;
using System;
using System.Linq;
using System.Linq.Expressions;

namespace GameStoreAPI.Queries
{
    public class GameListQuery
    {
        public SortDirection SortDirection { get; set; }
        public GameListSortField? SortField { get; set; }
        public long? Page { get; set; }
        public long? PageSize { get; set; }
        public string TitleLike { get; set; }
        public DateTime? ReleaseDateFrom { get; set; }
        public DateTime? ReleaseDateTo { get; set; }
        public decimal? ScoreFrom { get; set; }
        public decimal? ScoreTo { get; set; }
        public decimal? PriceFrom { get; set; }
        public decimal? PriceTo { get; set; }
        public long? PlatformId { get; set; }
        public long? PublisherId { get; set; }
        public long? DeveloperId { get; set; }
        public long? GenreId { get; set; }
    }

    public class GameListSortField
    {
        public static readonly GameListSortField Title = new GameListSortField("title");
        public static readonly GameListSortField ReleaseDate = new GameListSortField("releaseDate");
        public static readonly GameListSortField Score = new GameListSortField("score");
        public static readonly GameListSortField Price = new GameListSortField("price");

        private GameListSortField(string field)
        {
            Field = field;
        }

        public string Field { get; private set; }

        public static GameListSortField FromString(string field)
        {
            if (field == null)
            {
                throw new ArgumentNullException(nameof(field));
            }

            if (field.Equals(Title.Field, StringComparison.OrdinalIgnoreCase))
            {
                return Title;
            }

            if (field.Equals(ReleaseDate.Field, StringComparison.OrdinalIgnoreCase))
            {
                return ReleaseDate;
            }

            if (field.Equals(Score.Field, StringComparison.OrdinalIgnoreCase))
            {
                return Score;
            }

            if (field.Equals(Price.Field, StringComparison.OrdinalIgnoreCase))
            {
                return Price;
            }

            throw new ArgumentException($"Invalid sort field: {field}");
        }
    }

    public class GameListQuerySortField<T>
    {
        public static readonly GameListQuerySortField<string> Title = new GameListQuerySortField<string>(g => g.Title);
        public static readonly GameListQuerySortField<DateTime> ReleaseDate = new GameListQuerySortField<DateTime>(g => g.ReleaseDate);
        public static readonly GameListQuerySortField<decimal> Score = new GameListQuerySortField<decimal>(g => g.Score);
        public static readonly GameListQuerySortField<decimal> Price = new GameListQuerySortField<decimal>(g => g.Price);

        private readonly Expression<Func<Game, T>> keySelector;

        public Expression KeySelector { get => keySelector; }

        private GameListQuerySortField(Expression<Func<Game, T>> keySelector)
        {
            this.keySelector = keySelector;
        }
    }

    public static class GameListQueryExtensions
    {
        public static IQueryable<Game> UseQuery(this IQueryable<Game> games, GameListQuery query)
        {
            if (query.SortDirection != null && query.SortField != null)
            {
                if (query.SortDirection == SortDirection.Ascending)
                {
                    if (query.SortField == GameListSortField.Title)
                    {
                        games = games.OrderBy(g => g.Title).ThenBy(g => g.Id);
                    }
                    else if (query.SortField == GameListSortField.ReleaseDate)
                    {
                        games = games.OrderBy(g => g.ReleaseDate).ThenBy(g => g.Id);
                    }
                    else if (query.SortField == GameListSortField.Score)
                    {
                        games = games.OrderBy(g => g.Score).ThenBy(g => g.Id);
                    }
                    else if (query.SortField == GameListSortField.Price)
                    {
                        games = games.OrderBy(g => g.Price).ThenBy(g => g.Id);
                    }
                }
                else
                {
                    if (query.SortField == GameListSortField.Title)
                    {
                        games = games.OrderByDescending(g => g.Title).ThenBy(g => g.Id);
                    }
                    else if (query.SortField == GameListSortField.ReleaseDate)
                    {
                        games = games.OrderByDescending(g => g.ReleaseDate).ThenBy(g => g.Id);
                    }
                    else if (query.SortField == GameListSortField.Score)
                    {
                        games = games.OrderByDescending(g => g.Score).ThenBy(g => g.Id);
                    }
                    else if (query.SortField == GameListSortField.Price)
                    {
                        games = games.OrderByDescending(g => g.Price).ThenBy(g => g.Id);
                    }
                }
            }

            if (query.Page != null && query.PageSize != null)
            {
                games = games.Skip((int)(query.Page.Value * query.PageSize.Value));
            }

            if (query.TitleLike != null)
            {
                games = games.Where(g => g.Title.Contains(query.TitleLike));
            }

            if (query.ReleaseDateFrom != null)
            {
                games = games.Where(g => g.ReleaseDate >= query.ReleaseDateFrom);
            }

            if (query.ReleaseDateTo != null)
            {
                games = games.Where(g => g.ReleaseDate <= query.ReleaseDateTo);
            }

            if (query.ScoreFrom != null)
            {
                games = games.Where(g => g.Score >= query.ScoreFrom);
            }

            if (query.ScoreTo != null)
            {
                games = games.Where(g => g.Score <= query.ScoreTo);
            }

            if (query.PriceFrom != null)
            {
                games = games.Where(g => g.Price >= query.PriceFrom);
            }

            if (query.PriceTo != null)
            {
                games = games.Where(g => g.Price <= query.PriceTo);
            }

            if (query.PlatformId != null)
            {
                games = games.Where(g => g.PlatformId == query.PlatformId);
            }

            if (query.PublisherId != null)
            {
                games = games.Where(g => g.PublisherId == query.PublisherId);
            }

            if (query.DeveloperId != null)
            {
                games = games.Where(g => g.DeveloperId == query.DeveloperId);
            }

            if (query.GenreId != null)
            {
                games = games.Where(g => g.Genres.Any(g => g.Id == query.GenreId));
            }

            return games;
        }
    }
}

File: frontend/src/app/game-list/models/game-list-query.ts
import { HttpParams } from "@angular/common/http";

type Nullable<T> = T | null;
class DatePickerValue {
  constructor(public year: number, public month: number, public day: number) { }
};

export class GameListQuery {
  constructor(
    public sortDirection: Nullable<string> = "asc",
    public sortField: Nullable<string> = "title",
    public page: Nullable<number> = 0,
    public pageSize: Nullable<number> = 20,
    public titleLike: Nullable<string> = null,
    public releaseDateFrom: Nullable<DatePickerValue> = null,
    public releaseDateTo: Nullable<DatePickerValue> = null,
    public scoreFrom: Nullable<number> = null,
    public scoreTo: Nullable<number> = null,
    public priceFrom: Nullable<number> = null,
    public priceTo: Nullable<number> = null,
    public platformId: Nullable<number | string> = null,
    public publisherId: Nullable<number | string> = null,
    public developerId: Nullable<number | string> = null,
    public genreId: Nullable<number | string> = null
  ) { }

  resetToDefaults() {
    this.sortDirection = "asc";
    this.sortField = "title";
    this.page = 0;
    this.pageSize = 20;
    this.titleLike = null;
    this.releaseDateFrom = null;
    this.releaseDateTo = null;
    this.scoreFrom = null;
    this.scoreTo = null;
    this.priceFrom = null;
    this.priceTo = null;
    this.platformId = null;
    this.publisherId = null;
    this.developerId = null;
    this.genreId = null;
  }

  cleanFilters() {
    this.titleLike = null;
    this.releaseDateFrom = null;
    this.releaseDateTo = null;
    this.scoreFrom = null;
    this.scoreTo = null;
    this.priceFrom = null;
    this.priceTo = null;
    this.platformId = null;
    this.publisherId = null;
    this.developerId = null;
    this.genreId = null;
  }

  cleanFilter(filter: string) {
    switch (filter) {
      case "titleLike":
        this.titleLike = null;
        break;
      case "releaseDateFrom":
        this.releaseDateFrom = null;
        break;
      case "releaseDateTo":
        this.releaseDateTo = null;
        break;
      case "scoreFrom":
        this.scoreFrom = null;
        break;
      case "scoreTo":
        this.scoreTo = null;
        break;
      case "priceFrom":
        this.priceFrom = null;
        break;
      case "priceTo":
        this.priceTo = null;
        break;
      case "platformId":
        this.platformId = null;
        break;
      case "publisherId":
        this.publisherId = null;
        break;
      case "developerId":
        this.developerId = null;
        break;
      case "genreId":
        this.genreId = null;
        break;
      default:
        break;
    }
  }

  toHttpParams(): HttpParams {
    let params = new HttpParams();

    console.log(this);

    if (this.sortDirection) {
      params = params.set("sortDirection", this.sortDirection);
    }

    if (this.sortField) {
      params = params.set("sortField", this.sortField);
    }

    if (this.page) {
      params = params.set("page", this.page.toString());
    }

    if (this.pageSize) {
      params = params.set("pageSize", this.pageSize.toString());
    }

    if (this.titleLike) {
      params = params.set("titleLike", this.titleLike);
    }

    if (this.releaseDateFrom) {
      let dateString = `${this.releaseDateFrom.year}-${this.releaseDateFrom.month}-${this.releaseDateFrom.day}`;
      params = params.set("releaseDateFrom", dateString);
    }

    if (this.releaseDateTo) {
      let dateString = `${this.releaseDateTo.year}-${this.releaseDateTo.month}-${this.releaseDateTo.day}`;
      params = params.set("releaseDateTo", dateString);
    }

    if (this.scoreFrom) {
      params = params.set("scoreFrom", this.scoreFrom.toString());
    }

    if (this.scoreTo) {
      params = params.set("scoreTo", this.scoreTo.toString());
    }

    if (this.priceFrom) {
      params = params.set("priceFrom", this.priceFrom.toString());
    }

    if (this.priceTo) {
      params = params.set("priceTo", this.priceTo.toString());
    }

    if (this.platformId && this.platformId !== "null") {
      params = params.set("platformId", this.platformId.toString());
    }

    if (this.publisherId && this.publisherId !== "null") {
      params = params.set("publisherId", this.publisherId.toString());
    }

    if (this.developerId && this.developerId !== "null") {
      params = params.set("developerId", this.developerId.toString());
    }

    if (this.genreId && this.genreId !== "null") {
      params = params.set("genreId", this.genreId.toString());
    }

    return params;
  }
}
